core/trino-main/src/test/java/io/trino/execution/scheduler/TestEventDrivenTaskSource.java

```diff
@@ -0,0 +1,1031 @@
+                handleSources.add(handleSource);
+                exchanges.put(fragmentId, new TestingExchange(handleSource));
+            }
+        });
+
+        Map<PlanNodeId, SplitSource> splitSources = new HashMap<>();
+        Multimaps.asMap(splits).forEach(((planNodeId, connectorSplits) -> splitSources.put(planNodeId, new TestingSplitSource(executor, connectorSplits))));
+        splitSources.putAll(failingSplitSources);
+
+        EventDrivenTaskSource.Callback taskSourceCallback = failingCallback.orElse(new TestingTaskSourceCallback());
+        int partitionCount = getPartitionCount(sourceHandles.values(), splits.values());
+        FaultTolerantPartitioningScheme partitioningScheme = createPartitioningScheme(partitionCount);
+        AtomicLong getSplitInvocations = new AtomicLong();
+        Set<PlanNodeId> allSources = ImmutableSet.<PlanNodeId>builder()
+                .addAll(remoteSources.values())
+                .addAll(splits.keySet())
+                .build();
+        List<TaskDescriptor> taskDescriptors = null;
+        RuntimeException failure = null;
+        TestingSplitAssigner testingSplitAssigner = new TestingSplitAssigner(allSources);
+        try (EventDrivenTaskSource taskSource = new EventDrivenTaskSource(
+                exchanges,
+                remoteSources,
+                () -> splitSources,
+                failingSplitAssigner.orElse(testingSplitAssigner),
+                taskSourceCallback,
+                executor,
+                1,
+                1,
+                partitioningScheme,
+                (getSplitDuration) -> getSplitInvocations.incrementAndGet())) {
+            taskSource.start();
+            try {
+                if (taskSourceCallback instanceof FailingTaskSourceCallback callback) {
+                    taskDescriptors = callback.getTaskDescriptors().get(1, MINUTES);
+                }
+                else if (taskSourceCallback instanceof TestingTaskSourceCallback callback) {
+                    taskDescriptors = callback.getTaskDescriptorsFuture().get(1, MINUTES);
+                }
+                else {
+                    fail("unexpected callback: " + taskSourceCallback.getClass());
+                }
+                assertTrue(testingSplitAssigner.isFinished());
+            }
+            catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                failure = new RuntimeException(e);
+            }
+            catch (ExecutionException e) {
+                if (e.getCause() instanceof RuntimeException runtimeException) {
+                    failure = runtimeException;
+                }
+                else {
+                    failure = new RuntimeException(e);
+                }
+            }
+            catch (TimeoutException e) {
+                failure = new UncheckedTimeoutException(e);
+            }
+        }
+
+        for (ExchangeSourceHandleSource handleSource : handleSources) {
```

Comment:
bail out quickly if `failure != null`?

Reply:
These checks are important. I'm trying to verify that sources are getting closed in case of a failure.